PopPUNK/lineage_clustering.py

```diff
@@ -0,0 +1,680 @@
+            should be added.
+        use_accessory (bool)
+            Option to use accessory distances rather than core distances.
+        num_processes (int)
+            Number of CPUs to use for calculations.
+            
+    Returns:
+        combined (dict)
+            Assignment of each isolate to clusters by all ranks used.
+    """
+    
+    # process distance matrix
+    # - this should be sorted (PyTorch allows this to be done on GPUs)
+    # - then the functions can be reimplemented to run faster on a
+    #   sorted list
+    distance_index = 1 if use_accessory else 0
+    distances = X[:,distance_index]
+    
+    # sort distances
+    distance_ranks = np.argsort(distances)
+    distances = distances[distance_ranks]
+    
+    # determine whether ref-ref or ref-query analysis
+    isolate_list = rlist
+    
+    # determine whether novel analysis or modifying existing analysis
+    use_existing = False
+    neighbours = {}
+    lineage_seed = {}
+    lineage_clustering = {}
+    previous_lineage_clustering = {}
+    
+    null_cluster_value = len(isolate_list) + 1
+    for R in rank_list:
+        lineage_clustering[R] = {i:null_cluster_value for i in isolate_list}
+        lineage_seed[R] = {}
+        neighbours[R] = {}
+        previous_lineage_clustering[R] = {}
+    
+    # shared memory data structures
+    shm_distances = shared_memory.SharedMemory(create = True, size = distances.nbytes, name = 'shm_distances')
+    distances_shared = np.ndarray(distances.shape, dtype = distances.dtype, buffer = shm_distances.buf)
+    distances_shared[:] = distances[:]
+
+    shm_distance_ranks = shared_memory.SharedMemory(create = True, size = distance_ranks.nbytes, name = 'shm_distance_ranks')
+    distance_ranks_shared = np.ndarray(distance_ranks.shape, dtype = distance_ranks.dtype, buffer = shm_distance_ranks.buf)
+    distance_ranks_shared[:] = distance_ranks[:]
+
+    isolate_list_shared = shared_memory.ShareableList(isolate_list, name = 'shm_isolate_list')
+    
+    # run clustering for an individual R
+    if num_processes == 1:
+        for R in rank_list:
+            lineage_clustering[R], lineage_seed[R], neighbours[R], previous_lineage_clustering[R] = run_clustering_for_R(R,
+                                                                                        null_cluster_value = null_cluster_value,
+                                                                                        qlist = qlist,
+                                                                                        existing_scheme = existing_scheme,
+                                                                                        distances_length = distances.shape,
+                                                                                        distances_type = distances.dtype)
+
+
+    else:
+
+        with Pool(processes = num_processes) as pool:
+            results = pool.map(partial(run_clustering_for_R,
```

Comment:
Pass the shared objects here, rather than recreating them again in the function

Reply:
Done in da5b6df826fe1cafe622094f82cb8d3677354661.